### ISSUE TITLE
socks5 not working

### ISSUE THREAD

**Issue by gboisec** [rapid7/metasploit-framework#11513]
Hey currently I'm trying to Pivot through a Linux Server (in my virtual Lab) which I compromised. Now I wanted to use socks5 to reach another Webserver on the internal Infrastructure, but nothing happens. When I use socks4a with the same options it works pefectly.

Edit: Used ssh -D to open a new Socks5 Proxy and it works, so it seems to be really something strange in the msf5 module

Steps to reproduce

1. Compromise a Linux Host
2. run post module autoroute
3. run socks5 module
4. set up firefox socks5 with dns settings
5. trying to reach the new webpage but nothing happens (DNS and IP)

This section should also tell us any relevant information about the
environment; for example, if an exploit that used to work is failing,
tell us the victim operating system and service versions.

msfv5.0.9 (tried also with 5.0.8)
- [x ] Kali package via apt

OS = Kali Linux 2019.1

**Comment by jeffmcjunkin** [comment]
I've experienced the same issue on the latest Metasploit version.

Below, socks4a is running on port 4000, socks5 is running on port 5000.
```
msf5 auxiliary(server/socks4a) > route print

IPv4 Active Routing Table
=========================

   Subnet             Netmask            Gateway
   ------             -------            -------
   10.5.80.0          255.255.255.0      Session 2

[*] There are currently no IPv6 routes defined.
msf5 auxiliary(server/socks4a) >
[*] Starting the socks4a proxy server

msf5 auxiliary(server/socks4a) > jobs -v

Jobs
====

  Id  Name                       Payload  Payload opts  URIPATH  Start Time                 Handler opts  Persist
  --  ----                       -------  ------------  -------  ----------                 ------------  -------
  3   Auxiliary: server/socks5                                   2019-07-06 19:03:30 +0000                false
  4   Auxiliary: server/socks4a                                  2019-07-06 19:03:43 +0000                false

msf5 auxiliary(server/socks4a) > version
Framework: 5.0.35-dev-
Console  : 5.0.35-dev-
```

Now, some testing, first with socks4:
```
root@slingshot:~# curl -vvv --socks4 127.0.0.1:4000 http://10.5.80.1:80/wordpres                                      s/ | grep '<title>'
*   Trying 127.0.0.1...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*                                       SOCKS4 communication to 10.5.80.1:80
* SOCKS4 connect to 10.5.80.1 (locally resolved)
* SOCKS4 request granted.
* Connected to 127.0.0.1 (127.0.0.1) port 4000 (#0)
  0     0    0     0    0     0      0      0 --:--:--  0:00:05 --:--:--     0> GET /wordpress/ HTTP/1.1
> Host: 10.5.80.1
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Sat, 06 Jul 2019 19:05:46 GMT
< Server: Apache
< X-Frame-Options: SAMEORIGIN
< X-Powered-By: PHP/7.3.6
< Link: <http://10.5.80.1/wordpress/wp-json/>; rel="https://api.w.org/"
< Vary: Accept-Encoding
< Transfer-Encoding: chunked
< Content-Type: text/html; charset=UTF-8
<
{ [8644 bytes data]
        <title>SEC580&#039;s In-Class Blog! &#8211; Just another WordPress site</title>
100 12434    0 12434    0     0   2371      0 --:--:--  0:00:05 --:--:--  2371
* Connection #0 to host 10.5.80.1 left intact
```

Success!

And now, with socks5, attempted with `curl` and `proxychains4`:
```
root@slingshot:~# curl --help | grep socks5
     --socks5 HOST[:PORT]  SOCKS5 proxy on given host + port
     --socks5-hostname HOST[:PORT]  SOCKS5 proxy, pass host name to proxy
     --socks5-gssapi-service NAME  SOCKS5 proxy service name for GSS-API
     --socks5-gssapi-nec  Compatibility with NEC SOCKS5 server
root@slingshot:~# curl -vvv --socks5 127.0.0.1:5000 http://10.5.80.1:80/wordpress/ | grep '<title>'
*   Trying 127.0.0.1...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* 10
* 5
* 80
* 1
* Failed to receive SOCKS5 connect request ack.
* Closing connection 0
curl: (7) Failed to receive SOCKS5 connect request ack.
root@slingshot:~# curl -vvv --socks5-hostname 127.0.0.1:5000 http://10.5.80.1:80/wordpress/ | grep '<title>'
*   Trying 127.0.0.1...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Failed to receive SOCKS5 connect request ack.
* Closing connection 0
curl: (7) Failed to receive SOCKS5 connect request ack.

root@slingshot:~# tail -n 2 /etc/proxychains.conf
socks5 127.0.0.1 5000
root@slingshot:~# proxychains4 curl -vvv  http://10.5.80.1:80/wordpress/ | grep '<title>'
[proxychains] config file found: /etc/proxychains.conf
[proxychains] preloading /usr/local/lib/libproxychains4.so
[proxychains] DLL init: proxychains-ng 4.14
*   Trying 10.5.80.1...
[proxychains] Strict chain  ...  127.0.0.1:5000  ...  10.5.80.1:80 <--socket error or timeout!
* Immediate connect fail for 10.5.80.1: Connection refused
* Closing connection 0
curl: (7) Couldn't connect to server
```

**Comment by mubix** [comment]
ping :)

**Comment by zeroSteiner** [comment]
Proxying over sessions should work now that #12927 has been landed. Would someone be able to confirm that the bug I found and fixed addressed this?

**Comment by zeroSteiner** [comment]
The changes submitted in #12984 will bring the SOCKS 5 server's response to CONNECT requests in line with the [RFC's description](https://tools.ietf.org/html/rfc1928) by reporting the port to which the socket is bound on the system running meterpreter.

>    In the reply to a CONNECT, BND.PORT contains the port number that the
>    server assigned to connect to the target host, while BND.ADDR
>    contains the associated IP address.  The supplied BND.ADDR is often
>    different from the IP address that the client uses to reach the SOCKS
>    server, since such servers are often multi-homed.  It is expected
>    that the SOCKS server will use DST.ADDR and DST.PORT, and the
>    client-side source address and port in evaluating the CONNECT
>    request.

**Comment by mubix** [comment]
When attempting to use Burp through the SOCKS5 proxy I get a "SOCKS server general failure" using the current master / lastet as of today b85cd9b682c121941b9f64d27d11f6d6433d4273

**Comment by zeroSteiner** [comment]
@mubix so I'm pretty sure you're running into two issues.
1. (Already resolved) the payloads gem wasn't updated to include the changes I added to fix this
2. (PR pending) the `#getlocalname` method for sockets backed by meterpreter channels that use reverse HTTP(S) channels was breaking on a call to `#getsockname` due to the local socket already being closed, presumably due to the request being processed after the client/meterpreter had closed it. With these changes I've tested the SOCKS5 server with reverse_tcp, reverse_http, and reverse_https on the Windows and Python platforms, all with success.

**Comment by zeroSteiner** [comment]
Alright, since the PR to fix the HTTP-specific variant of this issue has been landed (see: [rapid7/rex-socket#23](https://github.com/rapid7/rex-socket/pull/23)) I'm going to close this out as resolved. If you're still running into errors feel free to comment here or open a new issue (tag me in that case so I can take a look please!).

**Comment by jeffmcjunkin** [comment]
For posterity, I tested again today using the new `auxilliary/server/socks_proxy` with a `reverse_tcp` meterpreter and was successful with both SOCKS proxy versions 4a and 5. Hooray! </delayed>
